File: unidbg-api/src/main/java/com/vortexdbg/hook/whale/IWhale.java
package com.vortexdbg.hook.whale;

import com.vortexdbg.Symbol;
import com.vortexdbg.hook.IHook;
import com.vortexdbg.hook.InlineHook;
import com.vortexdbg.hook.ReplaceCallback;

public interface IWhale extends IHook, InlineHook {

    void inlineHookFunction(long address, ReplaceCallback callback);
    void inlineHookFunction(Symbol symbol, ReplaceCallback callback);

    void inlineHookFunction(long address, ReplaceCallback callback, boolean enablePostCall);
    @SuppressWarnings("unused")
    void inlineHookFunction(Symbol symbol, ReplaceCallback callback, boolean enablePostCall);

    /**
     * 当前对android无效，参考：https://github.com/asLody/whale/blob/master/whale/src/whale.cc，只支持苹果
     */
    @SuppressWarnings("unused")
    void importHookFunction(String symbol, ReplaceCallback callback);
    void importHookFunction(String symbol, ReplaceCallback callback, boolean enablePostCall);

}

File: vortexdbg-api/src/main/kotlin/com/vortexdbg/hook/whale/IWhale.kt
package com.vortexdbg.hook.whale

import com.vortexdbg.Symbol
import com.vortexdbg.hook.IHook
import com.vortexdbg.hook.InlineHook
import com.vortexdbg.hook.ReplaceCallback

interface IWhale : IHook, InlineHook {

    fun inlineHookFunction(address: Long, callback: ReplaceCallback)
    fun inlineHookFunction(symbol: Symbol, callback: ReplaceCallback)

    fun inlineHookFunction(address: Long, callback: ReplaceCallback, enablePostCall: Boolean)
    @Suppress("unused")
    fun inlineHookFunction(symbol: Symbol, callback: ReplaceCallback, enablePostCall: Boolean)

    /**
     * Hook an imported (PLT/GOT) symbol. Apple/Mach-O only; this has no effect
     * on Android, mirroring the upstream whale implementation
     * (see whale/src/whale.cc in github.com/asLody/whale).
     */
    @Suppress("unused")
    fun importHookFunction(symbol: String, callback: ReplaceCallback)
    fun importHookFunction(symbol: String, callback: ReplaceCallback, enablePostCall: Boolean)

}
